{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre-process average_attention file to meet the previous ADsex project data structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def filter_edges(patient_type):\n",
    "\n",
    "    df = pd.read_csv(f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}.csv')\n",
    "    filtered_df = df[df['EdgeType'] == 'Gene-Prot-Gene-Prot']\n",
    "    filtered_df.to_csv(f'./ROSMAP-analysis/fold_0/filtered_average_attention_{patient_type}.csv', index=False)\n",
    "\n",
    "\n",
    "filter_edges('AD')\n",
    "filter_edges('NOAD')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## P-Values Part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Survival numbers: [1, 3, 6, 10, 12, 14, 15, 18, 19, 20, 21, 23, 32, 33, 35, 36, 38, 46, 47, 48, 51, 53, 55, 59, 61, 63, 65, 66, 69, 70, 71, 72, 73, 74, 75, 78, 79, 80, 81, 83, 84, 85, 86, 88, 91, 93, 95, 96, 99, 102, 103, 105, 106, 107, 110, 112, 113, 116, 117, 118, 119, 121, 124, 127]\n",
      "Survival numbers: [2, 4, 5, 7, 8, 9, 11, 13, 16, 17, 22, 24, 25, 26, 27, 28, 29, 30, 31, 34, 37, 39, 40, 41, 42, 43, 44, 45, 49, 50, 52, 54, 56, 57, 58, 60, 62, 64, 67, 68, 76, 77, 82, 87, 89, 90, 92, 94, 97, 98, 100, 101, 104, 108, 109, 111, 114, 115, 120, 122, 123, 125, 126, 128]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "def get_survival_numbers(folder_path):\n",
    "    files = os.listdir(folder_path)\n",
    "    survival_numbers = []\n",
    "    for file in files:\n",
    "        match = re.match(r'survival(\\d+)\\.csv', file)\n",
    "        if match:\n",
    "            number = int(match.group(1))\n",
    "            survival_numbers.append(number)\n",
    "    return survival_numbers\n",
    "\n",
    "folder_path_AD = './ROSMAP-analysis/fold_0/AD'\n",
    "folder_path_NOAD = './ROSMAP-analysis/fold_0/NOAD'\n",
    "survival_numbers_AD = sorted(get_survival_numbers(folder_path_AD))\n",
    "survival_numbers_NOAD = sorted(get_survival_numbers(folder_path_NOAD))\n",
    "print(\"Survival numbers:\", survival_numbers_AD)\n",
    "print(\"Survival numbers:\", survival_numbers_NOAD)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Survival numbers: ['R1042011', 'R1218460', 'R1407047', 'R1660726', 'R1743384', 'R1924801', 'R2045909', 'R2237803', 'R2373966', 'R2455930', 'R2494273', 'R2543886', 'R2880377', 'R2881301', 'R3143439', 'R3176125', 'R3368249', 'R3811781', 'R3922205', 'R3978789', 'R4260171', 'R4527133', 'R4641987', 'R5234179', 'R5374583', 'R5541746', 'R5656511', 'R5739959', 'R5850046', 'R5907586', 'R5927382', 'R5973191', 'R6001210', 'R6058053', 'R6108690', 'R6231758', 'R6267541', 'R6292415', 'R6392007', 'R6665276', 'R6698302', 'R6743664', 'R6751504', 'R6765868', 'R6879714', 'R7143693', 'R7325259', 'R7483736', 'R7796947', 'R7911202', 'R7915228', 'R8044591', 'R8155560', 'R8293796', 'R8466373', 'R8759568', 'R8774534', 'R9072353', 'R9101940', 'R9127940', 'R9330569', 'R9419876', 'R9598418', 'R9904978']\n",
      "Survival numbers: ['R1105988', 'R1262106', 'R1287407', 'R1489314', 'R1531359', 'R1617674', 'R1710143', 'R1822146', 'R2079629', 'R2111091', 'R2516394', 'R2575548', 'R2624931', 'R2678902', 'R2716798', 'R2721311', 'R2731324', 'R2731764', 'R2803490', 'R3008520', 'R3257830', 'R3408213', 'R3477250', 'R3535957', 'R3739042', 'R3740754', 'R3741788', 'R3757880', 'R4119160', 'R4147686', 'R4276053', 'R4531942', 'R4917253', 'R5177066', 'R5211056', 'R5369295', 'R5449861', 'R5546461', 'R5766881', 'R5789564', 'R6114572', 'R6132310', 'R6476704', 'R6759986', 'R6847727', 'R6862468', 'R7025378', 'R7295303', 'R7738727', 'R7791442', 'R7876628', 'R7881801', 'R7959406', 'R8444624', 'R8457289', 'R8724814', 'R8845676', 'R8937093', 'R9354381', 'R9489952', 'R9596785', 'R9680160', 'R9781891', 'R9936070']\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('./ROSMAP-graph-data/survival_label_map_dict.csv')\n",
    "\n",
    "def num_to_id(num):\n",
    "    return df.loc[df['individualID_Num'] == num, 'individualID'].values[0]\n",
    "\n",
    "\n",
    "survival_numbers_AD = [num_to_id(num) for num in survival_numbers_AD]\n",
    "survival_numbers_NOAD = [num_to_id(num) for num in survival_numbers_NOAD]\n",
    "print(\"Survival numbers:\", survival_numbers_AD)\n",
    "print(\"Survival numbers:\", survival_numbers_NOAD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AIFM2-Epi', 'ATM-Epi', 'ATR-Epi', 'CACNA2D2-Epi', 'CACNA2D4-Epi', 'CACNB4-Epi', 'CACNG1-Epi', 'CACNG5-Epi', 'CACNG6-Epi', 'CARD16-Epi', 'CARD17-Epi', 'CARD18-Epi', 'CASP1-Epi', 'CASP12-Epi', 'CBLB-Epi', 'CD72-Epi', 'CD8A-Epi', 'CD8B-Epi', 'CD8B2-Epi', 'CHEK1-Epi', 'CHEK2-Epi', 'CLEC4D-Epi', 'CLEC4E-Epi', 'CLEC6A-Epi', 'CTLA4-Epi', 'DDB2-Epi', 'FCER1G-Epi', 'GNAQ-Epi', 'GTSE1-Epi', 'IL17B-Epi', 'IL17C-Epi', 'IL17F-Epi', 'IL17RA-Epi', 'IL17RB-Epi', 'IL17RC-Epi', 'IL17RE-Epi', 'IL25-Epi', 'KCNJ12-Epi', 'KCNJ14-Epi', 'KCNJ18-Epi', 'KCNJ2-Epi', 'LCK-Epi', 'LILRA1-Epi', 'LILRA2-Epi', 'LILRA4-Epi', 'LILRA5-Epi', 'LILRA6-Epi', 'LILRB1-Epi', 'LILRB2-Epi', 'LILRB3-Epi', 'LILRB4-Epi', 'LILRB5-Epi', 'MYD88-Epi', 'OXTR-Epi', 'PDCD1-Epi', 'PERP-Epi', 'PMAIP1-Epi', 'PPM1D-Epi', 'PTPRC-Epi', 'RRM2-Epi', 'RRM2B-Epi', 'SERPINB5-Epi', 'STEAP3-Epi', 'TLR5-Epi', 'TLR7-Epi', 'TLR8-Epi', 'TLR9-Epi', 'TNFRSF10A-Epi', 'TNFRSF10B-Epi', 'TP53I3-Epi', 'AIFM2-Geno', 'ATM-Geno', 'ATR-Geno', 'CACNA2D2-Geno', 'CACNA2D4-Geno', 'CACNB4-Geno', 'CACNG1-Geno', 'CACNG5-Geno', 'CACNG6-Geno', 'CARD16-Geno', 'CARD17-Geno', 'CARD18-Geno', 'CASP1-Geno', 'CASP12-Geno', 'CBLB-Geno', 'CD72-Geno', 'CD8A-Geno', 'CD8B-Geno', 'CD8B2-Geno', 'CHEK1-Geno', 'CHEK2-Geno', 'CLEC4D-Geno', 'CLEC4E-Geno', 'CLEC6A-Geno', 'CTLA4-Geno', 'DDB2-Geno', 'FCER1G-Geno', 'GNAQ-Geno', 'GTSE1-Geno', 'IL17B-Geno', 'IL17C-Geno', 'IL17F-Geno', 'IL17RA-Geno', 'IL17RB-Geno', 'IL17RC-Geno', 'IL17RE-Geno', 'IL25-Geno', 'KCNJ12-Geno', 'KCNJ14-Geno', 'KCNJ18-Geno', 'KCNJ2-Geno', 'LCK-Geno', 'LILRA1-Geno', 'LILRA2-Geno', 'LILRA4-Geno', 'LILRA5-Geno', 'LILRA6-Geno', 'LILRB1-Geno', 'LILRB2-Geno', 'LILRB3-Geno', 'LILRB4-Geno', 'LILRB5-Geno', 'MYD88-Geno', 'OXTR-Geno', 'PDCD1-Geno', 'PERP-Geno', 'PMAIP1-Geno', 'PPM1D-Geno', 'PTPRC-Geno', 'RRM2-Geno', 'RRM2B-Geno', 'SERPINB5-Geno', 'STEAP3-Geno', 'TLR5-Geno', 'TLR7-Geno', 'TLR8-Geno', 'TLR9-Geno', 'TNFRSF10A-Geno', 'TNFRSF10B-Geno', 'TP53I3-Geno', 'AIFM2-Tran', 'ATM-Tran', 'ATR-Tran', 'CACNA2D2-Tran', 'CACNA2D4-Tran', 'CACNB4-Tran', 'CACNG1-Tran', 'CACNG5-Tran', 'CACNG6-Tran', 'CARD16-Tran', 'CARD17-Tran', 'CARD18-Tran', 'CASP1-Tran', 'CASP12-Tran', 'CBLB-Tran', 'CD72-Tran', 'CD8A-Tran', 'CD8B-Tran', 'CD8B2-Tran', 'CHEK1-Tran', 'CHEK2-Tran', 'CLEC4D-Tran', 'CLEC4E-Tran', 'CLEC6A-Tran', 'CTLA4-Tran', 'DDB2-Tran', 'FCER1G-Tran', 'GNAQ-Tran', 'GTSE1-Tran', 'IL17B-Tran', 'IL17C-Tran', 'IL17F-Tran', 'IL17RA-Tran', 'IL17RB-Tran', 'IL17RC-Tran', 'IL17RE-Tran', 'IL25-Tran', 'KCNJ12-Tran', 'KCNJ14-Tran', 'KCNJ18-Tran', 'KCNJ2-Tran', 'LCK-Tran', 'LILRA1-Tran', 'LILRA2-Tran', 'LILRA4-Tran', 'LILRA5-Tran', 'LILRA6-Tran', 'LILRB1-Tran', 'LILRB2-Tran', 'LILRB3-Tran', 'LILRB4-Tran', 'LILRB5-Tran', 'MYD88-Tran', 'OXTR-Tran', 'PDCD1-Tran', 'PERP-Tran', 'PMAIP1-Tran', 'PPM1D-Tran', 'PTPRC-Tran', 'RRM2-Tran', 'RRM2B-Tran', 'SERPINB5-Tran', 'STEAP3-Tran', 'TLR5-Tran', 'TLR7-Tran', 'TLR8-Tran', 'TLR9-Tran', 'TNFRSF10A-Tran', 'TNFRSF10B-Tran', 'TP53I3-Tran', 'AIFM2', 'ATM', 'ATR', 'CACNA2D2', 'CACNA2D4', 'CACNB4', 'CACNG1', 'CACNG5', 'CACNG6', 'CARD16', 'CARD17', 'CARD18', 'CASP1', 'CASP12', 'CBLB', 'CD72', 'CD8A', 'CD8B', 'CD8B2', 'CHEK1', 'CHEK2', 'CLEC4D', 'CLEC4E', 'CLEC6A', 'CTLA4', 'DDB2', 'FCER1G', 'GNAQ', 'GTSE1', 'IL17B', 'IL17C', 'IL17F', 'IL17RA', 'IL17RB', 'IL17RC', 'IL17RE', 'IL25', 'KCNJ12', 'KCNJ14', 'KCNJ18', 'KCNJ2', 'LCK', 'LILRA1', 'LILRA2', 'LILRA4', 'LILRA5', 'LILRA6', 'LILRB1', 'LILRB2', 'LILRB3', 'LILRB4', 'LILRB5', 'MYD88', 'OXTR', 'PDCD1', 'PERP', 'PMAIP1', 'PPM1D', 'PTPN6', 'PTPRC', 'RRM2', 'RRM2B', 'SERPINB5', 'STEAP3', 'TLR5', 'TLR7', 'TLR8', 'TLR9', 'TNFRSF10A', 'TNFRSF10B', 'TP53', 'TP53I3']\n",
      "['AIFM2-Epi', 'ATM-Epi', 'ATR-Epi', 'CACNA2D2-Epi', 'CACNA2D4-Epi', 'CACNB4-Epi', 'CACNG1-Epi', 'CACNG5-Epi', 'CACNG6-Epi', 'CARD16-Epi', 'CARD17-Epi', 'CARD18-Epi', 'CASP1-Epi', 'CASP12-Epi', 'CBLB-Epi', 'CD72-Epi', 'CD8A-Epi', 'CD8B-Epi', 'CD8B2-Epi', 'CHEK1-Epi', 'CHEK2-Epi', 'CLEC4D-Epi', 'CLEC4E-Epi', 'CLEC6A-Epi', 'CTLA4-Epi', 'DDB2-Epi', 'FCER1G-Epi', 'GNAQ-Epi', 'GTSE1-Epi', 'IL17B-Epi', 'IL17C-Epi', 'IL17F-Epi', 'IL17RA-Epi', 'IL17RB-Epi', 'IL17RC-Epi', 'IL17RE-Epi', 'IL25-Epi', 'KCNJ12-Epi', 'KCNJ14-Epi', 'KCNJ18-Epi', 'KCNJ2-Epi', 'LCK-Epi', 'LILRA1-Epi', 'LILRA2-Epi', 'LILRA4-Epi', 'LILRA5-Epi', 'LILRA6-Epi', 'LILRB1-Epi', 'LILRB2-Epi', 'LILRB3-Epi', 'LILRB4-Epi', 'LILRB5-Epi', 'MYD88-Epi', 'OXTR-Epi', 'PDCD1-Epi', 'PERP-Epi', 'PMAIP1-Epi', 'PPM1D-Epi', 'PTPRC-Epi', 'RRM2-Epi', 'RRM2B-Epi', 'SERPINB5-Epi', 'STEAP3-Epi', 'TLR5-Epi', 'TLR7-Epi', 'TLR8-Epi', 'TLR9-Epi', 'TNFRSF10A-Epi', 'TNFRSF10B-Epi', 'TP53I3-Epi', 'AIFM2-Geno', 'ATM-Geno', 'ATR-Geno', 'CACNA2D2-Geno', 'CACNA2D4-Geno', 'CACNB4-Geno', 'CACNG1-Geno', 'CACNG5-Geno', 'CACNG6-Geno', 'CARD16-Geno', 'CARD17-Geno', 'CARD18-Geno', 'CASP1-Geno', 'CASP12-Geno', 'CBLB-Geno', 'CD72-Geno', 'CD8A-Geno', 'CD8B-Geno', 'CD8B2-Geno', 'CHEK1-Geno', 'CHEK2-Geno', 'CLEC4D-Geno', 'CLEC4E-Geno', 'CLEC6A-Geno', 'CTLA4-Geno', 'DDB2-Geno', 'FCER1G-Geno', 'GNAQ-Geno', 'GTSE1-Geno', 'IL17B-Geno', 'IL17C-Geno', 'IL17F-Geno', 'IL17RA-Geno', 'IL17RB-Geno', 'IL17RC-Geno', 'IL17RE-Geno', 'IL25-Geno', 'KCNJ12-Geno', 'KCNJ14-Geno', 'KCNJ18-Geno', 'KCNJ2-Geno', 'LCK-Geno', 'LILRA1-Geno', 'LILRA2-Geno', 'LILRA4-Geno', 'LILRA5-Geno', 'LILRA6-Geno', 'LILRB1-Geno', 'LILRB2-Geno', 'LILRB3-Geno', 'LILRB4-Geno', 'LILRB5-Geno', 'MYD88-Geno', 'OXTR-Geno', 'PDCD1-Geno', 'PERP-Geno', 'PMAIP1-Geno', 'PPM1D-Geno', 'PTPRC-Geno', 'RRM2-Geno', 'RRM2B-Geno', 'SERPINB5-Geno', 'STEAP3-Geno', 'TLR5-Geno', 'TLR7-Geno', 'TLR8-Geno', 'TLR9-Geno', 'TNFRSF10A-Geno', 'TNFRSF10B-Geno', 'TP53I3-Geno', 'AIFM2-Tran', 'ATM-Tran', 'ATR-Tran', 'CACNA2D2-Tran', 'CACNA2D4-Tran', 'CACNB4-Tran', 'CACNG1-Tran', 'CACNG5-Tran', 'CACNG6-Tran', 'CARD16-Tran', 'CARD17-Tran', 'CARD18-Tran', 'CASP1-Tran', 'CASP12-Tran', 'CBLB-Tran', 'CD72-Tran', 'CD8A-Tran', 'CD8B-Tran', 'CD8B2-Tran', 'CHEK1-Tran', 'CHEK2-Tran', 'CLEC4D-Tran', 'CLEC4E-Tran', 'CLEC6A-Tran', 'CTLA4-Tran', 'DDB2-Tran', 'FCER1G-Tran', 'GNAQ-Tran', 'GTSE1-Tran', 'IL17B-Tran', 'IL17C-Tran', 'IL17F-Tran', 'IL17RA-Tran', 'IL17RB-Tran', 'IL17RC-Tran', 'IL17RE-Tran', 'IL25-Tran', 'KCNJ12-Tran', 'KCNJ14-Tran', 'KCNJ18-Tran', 'KCNJ2-Tran', 'LCK-Tran', 'LILRA1-Tran', 'LILRA2-Tran', 'LILRA4-Tran', 'LILRA5-Tran', 'LILRA6-Tran', 'LILRB1-Tran', 'LILRB2-Tran', 'LILRB3-Tran', 'LILRB4-Tran', 'LILRB5-Tran', 'MYD88-Tran', 'OXTR-Tran', 'PDCD1-Tran', 'PERP-Tran', 'PMAIP1-Tran', 'PPM1D-Tran', 'PTPRC-Tran', 'RRM2-Tran', 'RRM2B-Tran', 'SERPINB5-Tran', 'STEAP3-Tran', 'TLR5-Tran', 'TLR7-Tran', 'TLR8-Tran', 'TLR9-Tran', 'TNFRSF10A-Tran', 'TNFRSF10B-Tran', 'TP53I3-Tran', 'AIFM2', 'ATM', 'ATR', 'CACNA2D2', 'CACNA2D4', 'CACNB4', 'CACNG1', 'CACNG5', 'CACNG6', 'CARD16', 'CARD17', 'CARD18', 'CASP1', 'CASP12', 'CBLB', 'CD72', 'CD8A', 'CD8B', 'CD8B2', 'CHEK1', 'CHEK2', 'CLEC4D', 'CLEC4E', 'CLEC6A', 'CTLA4', 'DDB2', 'FCER1G', 'GNAQ', 'GTSE1', 'IL17B', 'IL17C', 'IL17F', 'IL17RA', 'IL17RB', 'IL17RC', 'IL17RE', 'IL25', 'KCNJ12', 'KCNJ14', 'KCNJ18', 'KCNJ2', 'LCK', 'LILRA1', 'LILRA2', 'LILRA4', 'LILRA5', 'LILRA6', 'LILRB1', 'LILRB2', 'LILRB3', 'LILRB4', 'LILRB5', 'MYD88', 'OXTR', 'PDCD1', 'PERP', 'PMAIP1', 'PPM1D', 'PTPN6', 'PTPRC', 'RRM2', 'RRM2B', 'SERPINB5', 'STEAP3', 'TLR5', 'TLR7', 'TLR8', 'TLR9', 'TNFRSF10A', 'TNFRSF10B', 'TP53', 'TP53I3']\n"
     ]
    }
   ],
   "source": [
    "gene_names_df = pd.read_csv('./ROSMAP-analysis/fold_0/AD_gene_names.csv')\n",
    "gene_names_AD = gene_names_df['Gene_name'].tolist()\n",
    "gene_names_AD = [name.replace('-Prot', '') for name in gene_names_AD]\n",
    "print(gene_names_AD)\n",
    "gene_names_df_NOAD = pd.read_csv('./ROSMAP-analysis/fold_0/NOAD_gene_names.csv')\n",
    "gene_names_NOAD = gene_names_df_NOAD['Gene_name'].tolist()\n",
    "gene_names_NOAD = [name.replace('-Prot', '') for name in gene_names_NOAD]\n",
    "print(gene_names_NOAD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-cnv_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-gene-expression_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Core-Promoter_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Distal-Promoter_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Downstream_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Proximal-Promoter_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Upstream_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-proteomics_AD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-cnv_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-gene-expression_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Core-Promoter_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Distal-Promoter_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Downstream_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Proximal-Promoter_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Upstream_NOAD.csv\n",
      "Filtered data saved to ./ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-proteomics_NOAD.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def process_data(file_path, gene_names, survival_numbers, output_path):\n",
    "    data = pd.read_csv(file_path)\n",
    "\n",
    "    columns_to_keep = [num for num in survival_numbers]\n",
    "    columns_to_keep = ['gene_name'] + columns_to_keep \n",
    "    filtered_data = data[columns_to_keep]\n",
    "\n",
    "    filtered_data = filtered_data[filtered_data['gene_name'].isin(gene_names)]\n",
    "\n",
    "    filtered_data.to_csv(output_path, index=False)\n",
    "    print(f\"Filtered data saved to {output_path}\")\n",
    "\n",
    "files_to_process = [\n",
    "    './ROSMAP-process/processed-genotype-cnv.csv',\n",
    "    './ROSMAP-process/processed-genotype-gene-expression.csv',\n",
    "    './ROSMAP-process/processed-genotype-methy-Core-Promoter.csv',\n",
    "    './ROSMAP-process/processed-genotype-methy-Distal-Promoter.csv',\n",
    "    './ROSMAP-process/processed-genotype-methy-Downstream.csv',\n",
    "    './ROSMAP-process/processed-genotype-methy-Proximal-Promoter.csv',\n",
    "    './ROSMAP-process/processed-genotype-methy-Upstream.csv',\n",
    "    './ROSMAP-process/processed-genotype-proteomics.csv',\n",
    "]\n",
    "\n",
    "output_dir_AD = './ROSMAP-analysis/node_analysis_processed/AD/'\n",
    "output_dir_NOAD = './ROSMAP-analysis/node_analysis_processed/NOAD/'\n",
    "# mkdir\n",
    "os.makedirs(output_dir_AD, exist_ok=True)\n",
    "os.makedirs(output_dir_NOAD, exist_ok=True)\n",
    "\n",
    "for file_path in files_to_process:\n",
    "    file_name = file_path.split('/')[-1].replace('.csv', '_AD.csv')\n",
    "    output_path = f\"{output_dir_AD}{file_name}\"\n",
    "    process_data(file_path, gene_names_AD, survival_numbers_AD, output_path)\n",
    "\n",
    "for file_path in files_to_process:\n",
    "    file_name = file_path.split('/')[-1].replace('.csv', '_NOAD.csv')\n",
    "    output_path = f\"{output_dir_NOAD}{file_name}\"\n",
    "    process_data(file_path, gene_names_NOAD, survival_numbers_NOAD, output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P-values calculated and saved for processed-genotype-cnv_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-gene-expression_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-methy-Core-Promoter_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-methy-Distal-Promoter_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-methy-Downstream_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-methy-Proximal-Promoter_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-methy-Upstream_pvalues.csv\n",
      "P-values calculated and saved for processed-genotype-proteomics_pvalues.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def calculate_pvalues(AD_file_path, NOAD_file_path):\n",
    "    AD_data = pd.read_csv(AD_file_path)\n",
    "    NOAD_data = pd.read_csv(NOAD_file_path)\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for gene in AD_data['gene_name']:\n",
    "        AD_row = AD_data[AD_data['gene_name'] == gene].iloc[:, 1:].values.flatten()\n",
    "        NOAD_row = NOAD_data[NOAD_data['gene_name'] == gene].iloc[:, 1:].values.flatten()\n",
    "        # for gene only in ad or noad\n",
    "        if len(AD_row) == 0 or len(NOAD_row) == 0:\n",
    "            print(f\"Skipping gene {gene}: One of the rows is empty\")\n",
    "            continue\n",
    "\n",
    "        # # If the gene data is only present in one file, assign a p-value of 1e-25\n",
    "        # if len(AD_row) == 0 or len(NOAD_row) == 0:\n",
    "        #     results.append({'gene_name': gene, 'p_value':1e-25})\n",
    "        #     print(f\"Including gene {gene} with missing data in one group; assigned p-value of 1e-25\")\n",
    "        #     continue\n",
    "\n",
    "        contingency_table = pd.crosstab(AD_row, NOAD_row)\n",
    "\n",
    "        chi2, p, dof, expected = chi2_contingency(contingency_table)\n",
    "\n",
    "\n",
    "        results.append({'gene_name': gene, 'p_value': p})\n",
    "\n",
    "    results_df = pd.DataFrame(results)\n",
    "    return results_df\n",
    "\n",
    "AD_files = [\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-cnv_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-gene-expression_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Core-Promoter_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Distal-Promoter_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Downstream_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Proximal-Promoter_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-methy-Upstream_AD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/AD/processed-genotype-proteomics_AD.csv'\n",
    "]\n",
    "\n",
    "NOAD_files = [\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-cnv_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-gene-expression_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Core-Promoter_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Distal-Promoter_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Downstream_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Proximal-Promoter_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-methy-Upstream_NOAD.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/NOAD/processed-genotype-proteomics_NOAD.csv'\n",
    "]\n",
    "\n",
    "# mkdir pvalues folder\n",
    "os.makedirs('./ROSMAP-analysis/node_analysis_processed/pvalues/', exist_ok=True)\n",
    "\n",
    "for ad_file, noad_file in zip(AD_files, NOAD_files):\n",
    "    results_df = calculate_pvalues(ad_file, noad_file)\n",
    "    file_name = ad_file.split('/')[-1].replace('_AD.csv', '_pvalues.csv')\n",
    "    results_df.to_csv(f'./ROSMAP-analysis/node_analysis_processed/pvalues/{file_name}', index=False)\n",
    "    print(f\"P-values calculated and saved for {file_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\2102755886.py:38: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  filtered_results = pd.concat([filtered_results, output_data], ignore_index=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Category</th>\n",
       "      <th>P_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CACNG6-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.110675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CD8B2-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.158003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CHEK1-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.139792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CHEK2-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.197207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CLEC4D-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.194052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>CTLA4-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.120279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>DDB2-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.172647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>GNAQ-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.192500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>GTSE1-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.036805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>IL25-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.061223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>LCK-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.007495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>LILRA6-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.071707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>LILRB2-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.084584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>LILRB4-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.057073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>PERP-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.145308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>RRM2-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.001328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>TLR9-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.079009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Gene-Tran</td>\n",
       "      <td>TP53I3-Tran</td>\n",
       "      <td>gene-expression</td>\n",
       "      <td>0.145056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Gene-Epi</td>\n",
       "      <td>KCNJ12-Epi</td>\n",
       "      <td>methy-Core-Promoter</td>\n",
       "      <td>0.028083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Gene-Epi</td>\n",
       "      <td>STEAP3-Epi</td>\n",
       "      <td>methy-Core-Promoter</td>\n",
       "      <td>0.163476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Gene-Epi</td>\n",
       "      <td>PMAIP1-Epi</td>\n",
       "      <td>methy-Proximal-Promoter</td>\n",
       "      <td>0.176720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Gene-Prot</td>\n",
       "      <td>AIFM2-Prot</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>0.158380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Gene-Prot</td>\n",
       "      <td>CASP1-Prot</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>0.198587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Gene-Prot</td>\n",
       "      <td>MYD88-Prot</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>0.057083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     NodeType    Gene_name                 Category   P_value\n",
       "0   Gene-Tran  CACNG6-Tran          gene-expression  0.110675\n",
       "1   Gene-Tran   CD8B2-Tran          gene-expression  0.158003\n",
       "2   Gene-Tran   CHEK1-Tran          gene-expression  0.139792\n",
       "3   Gene-Tran   CHEK2-Tran          gene-expression  0.197207\n",
       "4   Gene-Tran  CLEC4D-Tran          gene-expression  0.194052\n",
       "5   Gene-Tran   CTLA4-Tran          gene-expression  0.120279\n",
       "6   Gene-Tran    DDB2-Tran          gene-expression  0.172647\n",
       "7   Gene-Tran    GNAQ-Tran          gene-expression  0.192500\n",
       "8   Gene-Tran   GTSE1-Tran          gene-expression  0.036805\n",
       "9   Gene-Tran    IL25-Tran          gene-expression  0.061223\n",
       "10  Gene-Tran     LCK-Tran          gene-expression  0.007495\n",
       "11  Gene-Tran  LILRA6-Tran          gene-expression  0.071707\n",
       "12  Gene-Tran  LILRB2-Tran          gene-expression  0.084584\n",
       "13  Gene-Tran  LILRB4-Tran          gene-expression  0.057073\n",
       "14  Gene-Tran    PERP-Tran          gene-expression  0.145308\n",
       "15  Gene-Tran    RRM2-Tran          gene-expression  0.001328\n",
       "16  Gene-Tran    TLR9-Tran          gene-expression  0.079009\n",
       "17  Gene-Tran  TP53I3-Tran          gene-expression  0.145056\n",
       "18   Gene-Epi   KCNJ12-Epi      methy-Core-Promoter  0.028083\n",
       "19   Gene-Epi   STEAP3-Epi      methy-Core-Promoter  0.163476\n",
       "20   Gene-Epi   PMAIP1-Epi  methy-Proximal-Promoter  0.176720\n",
       "21  Gene-Prot   AIFM2-Prot               proteomics  0.158380\n",
       "22  Gene-Prot   CASP1-Prot               proteomics  0.198587\n",
       "23  Gene-Prot   MYD88-Prot               proteomics  0.057083"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pvalues_files = [\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-cnv_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-gene-expression_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Core-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Distal-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Downstream_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Proximal-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Upstream_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-proteomics_pvalues.csv'\n",
    "]\n",
    "\n",
    "node_type_mapping = {\n",
    "    'cnv': 'Geno',\n",
    "    'gene-expression': 'Tran',\n",
    "    'methy-Core-Promoter': 'Epi',\n",
    "    'methy-Distal-Promoter': 'Epi',\n",
    "    'methy-Downstream': 'Epi',\n",
    "    'methy-Proximal-Promoter': 'Epi',\n",
    "    'methy-Upstream': 'Epi',\n",
    "    'proteomics': 'Prot'\n",
    "}\n",
    "\n",
    "filtered_results = pd.DataFrame(columns=['NodeType', 'Gene_name', 'Category', 'P_value'])\n",
    "\n",
    "for file_path in pvalues_files:\n",
    "\n",
    "    category = file_path.split('-genotype-')[1].split('_pvalues')[0]\n",
    "    data = pd.read_csv(file_path)\n",
    "    significant_data = data[data['p_value'] < 0.2]\n",
    "    output_data = pd.DataFrame({\n",
    "        'NodeType': 'Gene-' + node_type_mapping[category],\n",
    "        'Gene_name': significant_data['gene_name'].astype(str) + '-' + node_type_mapping[category],\n",
    "        'Category': category,\n",
    "        'P_value': significant_data['p_value']\n",
    "    })\n",
    "    filtered_results = pd.concat([filtered_results, output_data], ignore_index=True)\n",
    "\n",
    "print(len(filtered_results))\n",
    "display(filtered_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./ROSMAP-analysis/p_values_100.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\407130103.py:37: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  filtered_results = pd.concat([filtered_results, output_data], ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pvalues_files = [\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-cnv_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-gene-expression_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Core-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Distal-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Downstream_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Proximal-Promoter_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-methy-Upstream_pvalues.csv',\n",
    "    './ROSMAP-analysis/node_analysis_processed/pvalues/processed-genotype-proteomics_pvalues.csv'\n",
    "]\n",
    "\n",
    "node_type_mapping = {\n",
    "    'cnv': 'Geno',\n",
    "    'gene-expression': 'Tran',\n",
    "    'methy-Core-Promoter': 'Epi',\n",
    "    'methy-Distal-Promoter': 'Epi',\n",
    "    'methy-Downstream': 'Epi',\n",
    "    'methy-Proximal-Promoter': 'Epi',\n",
    "    'methy-Upstream': 'Epi',\n",
    "    'proteomics': 'Prot'\n",
    "}\n",
    "\n",
    "filtered_results = pd.DataFrame(columns=['NodeType', 'Gene_name', 'Category', 'P_value'])\n",
    "\n",
    "for file_path in pvalues_files:\n",
    "    category = file_path.split('-genotype-')[1].split('_pvalues')[0]\n",
    "    data = pd.read_csv(file_path)\n",
    "    significant_data = data[data['p_value'] < 0.2]\n",
    "    output_data = pd.DataFrame({\n",
    "        'NodeType': 'Gene-' + node_type_mapping[category],\n",
    "        'Gene_name': significant_data['gene_name'].astype(str) + '-' + node_type_mapping[category],\n",
    "        'Category': category,\n",
    "        'P_value': significant_data['p_value']\n",
    "    })\n",
    "    filtered_results = pd.concat([filtered_results, output_data], ignore_index=True)\n",
    "\n",
    "filtered_results = filtered_results.sort_values(by='P_value', ascending=True)\n",
    "\n",
    "# top_values = [20, 30, 50, 70]\n",
    "top_values = [70]\n",
    "for top in top_values:\n",
    "    top_results = filtered_results.head(top)\n",
    "    filename = f'./ROSMAP-analysis/p_values_{top}.csv'\n",
    "    print(filename)\n",
    "    top_results.to_csv(filename, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# # top_values = [20, 30, 50, 70]\n",
    "# top_values = [20]\n",
    "\n",
    "# for top in top_values:\n",
    "#     df = pd.read_csv(f'./ROSMAP-analysis/p_values_{top}.csv')\n",
    "\n",
    "#     new_rows = []\n",
    "\n",
    "#     for index, row in df.iterrows():\n",
    "#         category = row['Category']\n",
    "        \n",
    "#         if category.startswith('cnv'):\n",
    "#             df.at[index, 'Category'] = 'mutation'\n",
    "#         elif category in ['gene-expression', 'methy-Core-Promoter', 'methy-Distal-Promoter',\n",
    "#                         'methy-Downstream', 'methy-Proximal-Promoter', 'methy-Upstream', 'proteomics']:\n",
    "#             new_row = row.copy()\n",
    "#             new_row['Category'] = 'mutation'\n",
    "#             new_rows.append(new_row)\n",
    "\n",
    "#     if new_rows:\n",
    "#         df = pd.concat([df, pd.DataFrame(new_rows)], ignore_index=True)\n",
    "\n",
    "#     df = df.drop_duplicates()\n",
    "\n",
    "#     df.to_csv(f'./ROSMAP-analysis/modified_p_values_{top}.csv', index=False)\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# top_values = [20, 30, 50, 70]\n",
    "top_values = [70]\n",
    "\n",
    "for top in top_values:\n",
    "    df = pd.read_csv(f'./ROSMAP-analysis/p_values_{top}.csv')\n",
    "\n",
    "    new_rows = []\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        category = row['Category']\n",
    "        \n",
    "        if category.startswith('cnv'):\n",
    "            df.at[index, 'Category'] = 'geno'\n",
    "        elif category in ['gene-expression', 'methy-Core-Promoter', 'methy-Distal-Promoter',\n",
    "                        'methy-Downstream', 'methy-Proximal-Promoter', 'methy-Upstream', 'proteomics']:\n",
    "            new_row = row.copy()\n",
    "            new_row['Category'] = 'geno'\n",
    "            new_row['NodeType'] = row['NodeType'].replace('Prot', 'Geno').replace('Tran', 'Geno').replace('Epi', 'Geno')\n",
    "            new_row['Gene_name'] = row['Gene_name'].replace('Prot', 'Geno').replace('Tran', 'Geno').replace('Epi', 'Geno')\n",
    "            new_rows.append(new_row)\n",
    "\n",
    "    if new_rows:\n",
    "        df = pd.concat([df, pd.DataFrame(new_rows)], ignore_index=True)\n",
    "\n",
    "    df = df.drop_duplicates()\n",
    "\n",
    "    df.to_csv(f'./ROSMAP-analysis/modified_p_values_{top}.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data saved for AD.\n",
      "Filtered data saved for NOAD.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def filter_attention_data(patient_type):\n",
    "\n",
    "    df = pd.read_csv(f'./ROSMAP-analysis/fold_0/filtered_average_attention_{patient_type}.csv')\n",
    "    \n",
    "    rows_to_keep = []\n",
    "    threshold = 0.15\n",
    "\n",
    "    for _, row in df.iterrows():\n",
    "        # Define conditions for keeping rows based on attention thresholds\n",
    "        forward_condition = (df['From'] == row['From']) & (df['To'] == row['To']) & (df['Attention'] > threshold)\n",
    "        backward_condition = (df['From'] == row['To']) & (df['To'] == row['From']) & (df['Attention'] > threshold)\n",
    "\n",
    "        # Check if any row meets the forward or backward conditions\n",
    "        if df[forward_condition].any(axis=0).any() or df[backward_condition].any(axis=0).any():\n",
    "            rows_to_keep.append(row)\n",
    "\n",
    "    result_df = pd.DataFrame(rows_to_keep).drop_duplicates()\n",
    "\n",
    "    result_df.to_csv(f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_ht.csv', index=False)\n",
    "    print(f\"Filtered data saved for {patient_type}.\")\n",
    "\n",
    "filter_attention_data('AD')\n",
    "filter_attention_data('NOAD')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cut the single branch (like peeling an onion, remove one single protein node at a time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved for AD with top 100.\n",
      "Data saved for NOAD with top 100.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def clean_attention_data(patient_type):\n",
    "    # top_values = [20, 30, 50, 70]\n",
    "    top_values = [70]\n",
    "\n",
    "    for top in top_values:\n",
    "        # Load gene number mapping data\n",
    "        map_all_gene = pd.read_csv('./ROSMAP-graph-data/map-all-gene.csv')\n",
    "        gene_name_to_num = map_all_gene.set_index('Gene_name')['Gene_num'].to_dict()\n",
    "\n",
    "        # Load p_values data and convert gene names\n",
    "        p_values = pd.read_csv(f'./ROSMAP-analysis/modified_p_values_{top}.csv')\n",
    "        important_gene_names = p_values['Gene_name'].apply(lambda x: x.split('-')[0] + '-Prot').unique()\n",
    "\n",
    "        # Get important gene numbers\n",
    "        important_gene_nums = set()\n",
    "        for gene_name in important_gene_names:\n",
    "            if gene_name in gene_name_to_num:\n",
    "                important_gene_nums.add(gene_name_to_num[gene_name])\n",
    "\n",
    "        # Load the connection data\n",
    "        df = pd.read_csv(f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_ht.csv')\n",
    "\n",
    "        changes = True\n",
    "        while changes:\n",
    "            # Recompute the number of unique 'To' for each 'From'\n",
    "            unique_links = df.groupby('From')['To'].nunique()\n",
    "            \n",
    "            # Apply the filter condition: remove rows where 'From' connects to a unique 'To' and does not involve important genes\n",
    "            initial_count = len(df)\n",
    "            to_remove = df[(df['From'].map(unique_links) == 1) &\n",
    "                        (~df['From'].isin(important_gene_nums)) &\n",
    "                        (~df['To'].isin(important_gene_nums))]\n",
    "\n",
    "            if not to_remove.empty:\n",
    "                # Create a reversed rows DataFrame\n",
    "                reversed_rows = to_remove[['To', 'From']].copy()\n",
    "                reversed_rows.columns = ['From', 'To']\n",
    "\n",
    "                # Check if the reversed rows exist in the original data\n",
    "                reversed_rows['is_reversed'] = reversed_rows.apply(lambda x: ((df['From'] == x['From']) & \n",
    "                                                                            (df['To'] == x['To']) ).any(), axis=1)\n",
    "\n",
    "                # Get the indices of all reversed rows marked as True\n",
    "                reversed_indices = df[df.apply(lambda x: ((reversed_rows['From'] == x['From']) & \n",
    "                                                        (reversed_rows['To'] == x['To']) & \n",
    "                                                        reversed_rows['is_reversed']).any(), axis=1)].index\n",
    "\n",
    "                # Merge the indices of the original and reversed rows\n",
    "                indices_to_remove = to_remove.index.union(reversed_indices)\n",
    "                df = df.drop(indices_to_remove)\n",
    "\n",
    "            changes = len(df) < initial_count\n",
    "\n",
    "        df.to_csv(f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_final_{top}.csv', index=False)\n",
    "        print(f\"Data saved for {patient_type} with top {top}.\")\n",
    "\n",
    "clean_attention_data('AD')\n",
    "clean_attention_data('NOAD')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Unique links (for debug only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# df = pd.read_csv('./fold_0/average_attention_AD_ht.csv')\n",
    "\n",
    "# unique_links = df.groupby('From')['To'].nunique()\n",
    "\n",
    "# unique_links_df = unique_links.reset_index(name='Unique Links')\n",
    "\n",
    "# result_df = pd.merge(df, unique_links_df, on='From', how='left')\n",
    "\n",
    "# result_df.to_csv('./fold_0/aaverage_attention_AD_ht_ul.csv', index=False)\n",
    "\n",
    "# df = pd.read_csv('./fold_0/average_attention_AD_final.csv')\n",
    "\n",
    "# unique_links = df.groupby('From')['To'].nunique()\n",
    "\n",
    "# unique_links_df = unique_links.reset_index(name='Unique Links')\n",
    "\n",
    "# result_df = pd.merge(df, unique_links_df, on='From', how='left')\n",
    "\n",
    "# result_df.to_csv('./fold_0/aaverage_attention_AD_final_ul.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the top 2 path between each hub gene nodes (prot nodes' unique link > 2) and remove the rest node\n",
    "## (NOT WORKING, will cause missing nodes when plotting)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Important gene numbers: {6660, 8461, 6797, 6798, 7577, 6809, 6941, 8222, 7197, 6690, 8486, 8368, 7602, 7604, 7606, 7996, 7749, 7500, 7897, 7261, 6884, 6500, 7413, 6775}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Attention_start'] = potential_paths['Attention_start']\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:47: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Attention_end'] = potential_paths['Attention_end']\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:48: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Mean Attention'] = (paths_df['Attention_start'] + paths_df['Attention_end']) / 2\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:83: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  filter_final_paths = final_paths.groupby(['From', 'To']).apply(lambda x: filter_paths(x, important_gene_nums)).reset_index(drop=True)\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Attention_start'] = potential_paths['Attention_start']\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:47: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Attention_end'] = potential_paths['Attention_end']\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:48: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  paths_df['Mean Attention'] = (paths_df['Attention_start'] + paths_df['Attention_end']) / 2\n",
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_21364\\3928268476.py:83: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  filter_final_paths = final_paths.groupby(['From', 'To']).apply(lambda x: filter_paths(x, important_gene_nums)).reset_index(drop=True)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def load_gene_information(top):\n",
    "    # Load p_values data and convert gene names\n",
    "    p_values = pd.read_csv(f'./ROSMAP-analysis/modified_p_values_{top}.csv')\n",
    "    important_gene_names = p_values['Gene_name'].apply(lambda x: x.split('-')[0] + '-Prot').unique()\n",
    "\n",
    "    # Load gene number mapping data\n",
    "    map_all_gene = pd.read_csv('./ROSMAP-graph-data/map-all-gene.csv')\n",
    "    gene_name_to_num = map_all_gene.set_index('Gene_name')['Gene_num'].to_dict()\n",
    "    \n",
    "    # Create reverse mapping from gene number to gene name\n",
    "    num_to_gene_name = {v: k for k, v in gene_name_to_num.items()}\n",
    "\n",
    "    # Get important gene numbers\n",
    "    important_gene_nums = set()\n",
    "    for gene_name in important_gene_names:\n",
    "        if gene_name in gene_name_to_num:\n",
    "            important_gene_nums.add(gene_name_to_num[gene_name])\n",
    "    print(\"Important gene numbers:\", important_gene_nums)\n",
    "    return important_gene_nums, num_to_gene_name\n",
    "\n",
    "\n",
    "def analyze_paths(patient_type, top, important_gene_nums):\n",
    "\n",
    "    df = pd.read_csv(f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_final_{top}.csv')\n",
    "\n",
    "    # Calculate unique links for each 'From'\n",
    "    unique_links = df.groupby('From')['To'].nunique()\n",
    "\n",
    "    # Identify nodes with more than 2 unique links\n",
    "    nodes_with_multiple_links = unique_links[unique_links > 2].index\n",
    "\n",
    "    # Identify all potential paths through any middle node\n",
    "    potential_starts = df[df['From'].isin(nodes_with_multiple_links)]\n",
    "    potential_ends = df[df['To'].isin(nodes_with_multiple_links)]\n",
    "\n",
    "    # Combine to find all possible paths through any middle nodes\n",
    "    potential_paths = pd.merge(potential_starts, potential_ends, left_on='To', right_on='From', suffixes=('_start', '_end'))\n",
    "\n",
    "    # Create a new DataFrame to display the path structure\n",
    "    paths_df = potential_paths[['From_start', 'To_start', 'To_end']]\n",
    "    paths_df.columns = ['From', 'Middle', 'To']\n",
    "\n",
    "    # Calculate the mean attention for each path\n",
    "    paths_df['Attention_start'] = potential_paths['Attention_start']\n",
    "    paths_df['Attention_end'] = potential_paths['Attention_end']\n",
    "    paths_df['Mean Attention'] = (paths_df['Attention_start'] + paths_df['Attention_end']) / 2\n",
    "\n",
    "    # Aggregate the data for identical paths to calculate mean values\n",
    "    aggregated_paths = paths_df.groupby(['From', 'To', 'Middle']).agg({\n",
    "        'Attention_start': 'mean',\n",
    "        'Attention_end': 'mean',\n",
    "        'Mean Attention': 'mean'\n",
    "    }).reset_index()\n",
    "\n",
    "    # Average the values for each path and its reverse path\n",
    "    aggregated_paths['Key'] = aggregated_paths.apply(lambda x: tuple(sorted([x['From'], x['To']])), axis=1)\n",
    "    grouped = aggregated_paths.groupby(['Key', 'Middle']).agg({\n",
    "        'Attention_start': 'mean',\n",
    "        'Attention_end': 'mean',\n",
    "        'Mean Attention': 'mean'\n",
    "    }).reset_index()\n",
    "\n",
    "    # Merge back to get the full structure with averaged values\n",
    "    final_paths = grouped.merge(aggregated_paths[['From', 'To', 'Middle', 'Key']], on=['Middle', 'Key'])\n",
    "    final_paths = final_paths.drop(columns=['Key'])\n",
    "\n",
    "    # Remove paths where From and To are the same\n",
    "    final_paths = final_paths[final_paths['From'] != final_paths['To']]\n",
    "\n",
    "    # Only keep paths where From < To to maintain a single direction per pair\n",
    "    final_paths = final_paths[final_paths['From'] < final_paths['To']].drop_duplicates()\n",
    "    \n",
    "    # Final path filtering based on important middle nodes and mean attention\n",
    "    def filter_paths(group, important_gene_nums):\n",
    "        # Filter paths ensuring that important paths are always retained\n",
    "        important_paths = group[group['Middle'].isin(important_gene_nums)]\n",
    "        top_paths = group.sort_values(by='Mean Attention', ascending=False).head(2)\n",
    "        combined_paths = pd.concat([important_paths, top_paths]).drop_duplicates().head(2)\n",
    "        return combined_paths\n",
    "\n",
    "    filter_final_paths = final_paths.groupby(['From', 'To']).apply(lambda x: filter_paths(x, important_gene_nums)).reset_index(drop=True)\n",
    "\n",
    "    # Record removable middle nodes, ensuring important nodes are not included\n",
    "    all_middles = set(final_paths['Middle'])\n",
    "    important_middles = set(final_paths[final_paths['Middle'].isin(important_gene_nums)]['Middle'])\n",
    "    removable_middles = all_middles - set(filter_final_paths['Middle']) - important_middles\n",
    "\n",
    "    \n",
    "    return filter_final_paths, removable_middles\n",
    "\n",
    "# top_values = [20, 30, 50, 70]\n",
    "top_values = [70]\n",
    "\n",
    "for top in top_values:\n",
    "    important_gene_nums, num_to_gene_name = load_gene_information(top)\n",
    "\n",
    "    for patient_type in ['AD', 'NOAD']:\n",
    "        final_paths_df, removable_middles = analyze_paths(patient_type, top, important_gene_nums)\n",
    "        removable_gene_names = {num_to_gene_name[node] for node in removable_middles if node in num_to_gene_name}\n",
    "        removable_gene_names_df = pd.DataFrame(removable_gene_names, columns=['Gene_name'])\n",
    "        removable_gene_names_df.to_csv(f'./ROSMAP-analysis/{patient_type}_middle_nodes_to_remove_{top}.csv', index=False)\n",
    "        # final_paths_df.to_csv(f'{patient_type}_paths_test_{top}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files for top 100 saved\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# top_values = [20, 30, 50, 70]\n",
    "top_values = [70]\n",
    "\n",
    "for top in top_values:\n",
    "    AD_data = pd.read_csv(f'./ROSMAP-analysis/AD_middle_nodes_to_remove_{top}.csv')\n",
    "    NOAD_data = pd.read_csv(f'./ROSMAP-analysis/NOAD_middle_nodes_to_remove_{top}.csv')\n",
    "\n",
    "    # Find the common rows\n",
    "    common_rows = pd.merge(AD_data, NOAD_data, on='Gene_name')\n",
    "\n",
    "    # Find the different rows\n",
    "    AD_diff = AD_data[~AD_data['Gene_name'].isin(NOAD_data['Gene_name'])]\n",
    "    NOAD_diff = NOAD_data[~NOAD_data['Gene_name'].isin(AD_data['Gene_name'])]\n",
    "\n",
    "    # Write the common rows to a new CSV file\n",
    "    common_rows.to_csv(f'./ROSMAP-analysis/middle_nodes_to_remove_{top}.csv', index=False)\n",
    "\n",
    "    # Write the different rows to a new CSV file\n",
    "    with open(f'./ROSMAP-analysis/middle_nodes_comparison_{top}.csv', 'w') as file:\n",
    "        file.write('AD:\\n')\n",
    "        AD_diff.to_csv(file, index=False)\n",
    "        file.write('\\nNOAD:\\n')\n",
    "        NOAD_diff.to_csv(file, index=False)\n",
    "\n",
    "    print(f\"Files for top {top} saved\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to ./ROSMAP-analysis/fold_0/average_attention_AD_final_cleaned_100.csv\n",
      "Saved to ./ROSMAP-analysis/fold_0/average_attention_NOAD_final_cleaned_100.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def clean_gene_connections(patient_type):\n",
    "\n",
    "    # top_values = [20, 30, 50, 70]\n",
    "    top_values = [70]\n",
    "    \n",
    "    for top in top_values:\n",
    "\n",
    "        # Load the gene number mapping data\n",
    "        map_all_gene = pd.read_csv('./ROSMAP-graph-data/map-all-gene.csv')\n",
    "        gene_name_to_num = map_all_gene.set_index('Gene_name')['Gene_num'].to_dict()\n",
    "\n",
    "        # Load the list of genes to remove with header now included\n",
    "        gene_list = pd.read_csv(f'./ROSMAP-analysis/middle_nodes_to_remove_{top}.csv')['Gene_name'].tolist()\n",
    "\n",
    "        # Convert names to numbers using dictionary comprehension\n",
    "        gene_numbers = {gene_name_to_num[gene] for gene in gene_list if gene in gene_name_to_num}\n",
    "\n",
    "        # Load the connection data\n",
    "        input_file = f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_final_{top}.csv'\n",
    "        df = pd.read_csv(input_file)\n",
    "\n",
    "        # Filter out all connections involving the genes in your list\n",
    "        filtered_df = df[~(df['From'].isin(gene_numbers) | df['To'].isin(gene_numbers))]\n",
    "\n",
    "        # Save the cleaned data to a new CSV file\n",
    "        output_file = f'./ROSMAP-analysis/fold_0/average_attention_{patient_type}_final_cleaned_{top}.csv'\n",
    "        filtered_df.to_csv(output_file, index=False)\n",
    "        print(f\"Saved to {output_file}\")\n",
    "\n",
    "clean_gene_connections('AD')\n",
    "clean_gene_connections('NOAD')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "def load_gene_information():\n",
    "    map_all_gene = pd.read_csv('./ROSMAP-graph-data/map-all-gene.csv')\n",
    "    gene_name_to_num = map_all_gene.set_index('Gene_name')['Gene_num'].to_dict()\n",
    "    num_to_gene_name = {num: name for name, num in gene_name_to_num.items()}\n",
    "    return gene_name_to_num, num_to_gene_name\n",
    "\n",
    "def process_file(filename, patient_type, num_to_gene_name, gene_nums):\n",
    "    df = pd.read_csv(filename)\n",
    "    average_attention_list = []\n",
    "\n",
    "    for num in gene_nums:\n",
    "        if num in num_to_gene_name:\n",
    "            relevant_rows = df[(df['From'] == num) | (df['To'] == num)]\n",
    "            average_attention = relevant_rows['Attention'].mean()\n",
    "            average_attention_list.append((num, average_attention))\n",
    "    \n",
    "    return average_attention_list\n",
    "\n",
    "def update_p_values_file(p_values_file, gene_name_to_num, num_to_gene_name):\n",
    "    p_values_df = pd.read_csv(p_values_file)\n",
    "\n",
    "    # Create a new column for modified Gene_name\n",
    "    p_values_df['Modified_Gene_name'] = p_values_df['Gene_name'].apply(lambda x: re.sub(r'-.*$', '-Prot', x))\n",
    "\n",
    "    # Map Modified_Gene_name to Gene_num\n",
    "    p_values_df['Gene_num'] = p_values_df['Modified_Gene_name'].map(gene_name_to_num)\n",
    "\n",
    "    # Drop rows where Gene_num is NaN (no mapping found)\n",
    "    p_values_df = p_values_df.dropna(subset=['Gene_num'])\n",
    "\n",
    "    gene_nums = p_values_df['Gene_num'].astype(int).tolist()\n",
    "    \n",
    "    # Process attention files for both patient types\n",
    "    _, num_to_gene_name = load_gene_information()\n",
    "    AD_attention = dict(process_file('./ROSMAP-analysis/fold_0/average_attention_AD.csv', 'AD', num_to_gene_name, gene_nums))\n",
    "    NOAD_attention = dict(process_file('./ROSMAP-analysis/fold_0/average_attention_NOAD.csv', 'NOAD', num_to_gene_name, gene_nums))\n",
    "\n",
    "    p_values_df['AD_Gene_weight'] = p_values_df['Gene_num'].apply(lambda x: AD_attention.get(x, 0))\n",
    "    p_values_df['NOAD_Gene_weight'] = p_values_df['Gene_num'].apply(lambda x: NOAD_attention.get(x, 0))\n",
    "\n",
    "    # Save to new CSV filea\n",
    "    p_values_df.to_csv('./ROSMAP-analysis/p_values_70_for_plot.csv', index=False)\n",
    "\n",
    "gene_name_to_num, num_to_gene_name = load_gene_information()\n",
    "update_p_values_file('./ROSMAP-analysis/p_values_70.csv', gene_name_to_num, num_to_gene_name)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "win_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
